error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]

### Description

When trying to build the project from source, it fails with the following error:

```yaml
error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
```

This seems to be caused because [glibc contains its own definition of `asprintf`](https://codebrowser.dev/glibc/glibc/libio/stdio.h.html#396) which has the `warn_unused_result` attribute.

If I remove `-Werror` and build the binary, everything works as expected and the `asprintf` linked
in the final binary is the one from your `utility.h`, so it's quite strange that `gcc` complains about it.

```console
$ objdump -T headsetcontrol | grep asprint
0000000000407650 g    DF .text	00000000000000b8  Base        asprintf
0000000000407510 g    DF .text	000000000000013c  Base        vasprintf
```

For completeness, here is the full build log:
```console
$ nix-shell -p cmake hidapi
$ mkdir build && cd build
$ cmake ..
$ make
[  3%] Building C object CMakeFiles/headsetcontrol.dir/src/main.c.o
/home/leix/Documents/github/HeadsetControl/src/main.c: In function ‘handle_feature’:
/home/leix/Documents/github/HeadsetControl/src/main.c:197:9: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  197 |         asprintf(&result.message, "This headset doesn't support %s", capabilities_str[cap]);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:208:13: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  208 |             asprintf(&result.message, "Could not open device. Error: %ls", hid_error(*device_handle));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:229:13: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  229 |             asprintf(&result.message, "Battery: %d%%", battery.level);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:247:17: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  247 |                 asprintf(&result.message, "Error retrieving battery status. Error: %ls", hid_error(*device_handle));
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:249:17: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  249 |                 asprintf(&result.message, "Error retrieving battery status");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:272:13: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  272 |             asprintf(&result.message, "Chat-Mix: %d", ret);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:327:9: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  327 |         asprintf(&result.message, "Failed to set/request %s, because of timeout", capabilities_str[cap]);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:330:9: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  330 |         asprintf(&result.message, "Failed to set/request %s. HeadsetControl Error", capabilities_str[cap]);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:333:9: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  333 |         asprintf(&result.message, "Failed to set/request %s. Provided parameter out of boundaries", capabilities_str[cap]);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:337:13: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  337 |             asprintf(&result.message, "Failed to set/request %s. Error: %d: %ls", capabilities_str[cap], ret, hid_error(*device_handle));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/leix/Documents/github/HeadsetControl/src/main.c:339:13: error: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  339 |             asprintf(&result.message, "Failed to set/request %s. Error: %d", capabilities_str[cap], ret);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/headsetcontrol.dir/build.make:76: CMakeFiles/headsetcontrol.dir/src/main.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:149: CMakeFiles/headsetcontrol.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

Here are some environment details:

```console
$ uname -a
Linux kuro 6.7.11 #1-NixOS SMP PREEMPT_DYNAMIC Tue Mar 26 22:22:50 UTC 2024 x86_64 GNU/Linux

$ cmake --version
cmake version 3.28.3

CMake suite maintained and supported by Kitware (kitware.com/cmake).

$ make --version
GNU Make 4.4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

$ gcc --version
gcc (GCC) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```



### On which OS does the problem happen?

Linux

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -36,7 +36,7 @@ IF (WIN32)
 	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-Bstatic,--whole-archive -lwinpthread -Wl,--no-whole-archive")
 ENDIF()
 
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Werror")
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall")
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
 # ------------------------------------------------------------------------------

--- a/src/main.c
+++ b/src/main.c
@@ -194,7 +194,7 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
     if ((device_found->capabilities & B(cap)) == 0) {
         result.status = FEATURE_ERROR;
         result.value  = -1;
-        asprintf(&result.message, "This headset doesn't support %s", capabilities_str[cap]);
+        _asprintf(&result.message, "This headset doesn't support %s", capabilities_str[cap]);
         return result;
     }
 
@@ -205,7 +205,7 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
         if (!device_handle | !(*device_handle)) {
             result.status = FEATURE_DEVICE_FAILED_OPEN;
             result.value  = 0;
-            asprintf(&result.message, "Could not open device. Error: %ls", hid_error(*device_handle));
+            _asprintf(&result.message, "Could not open device. Error: %ls", hid_error(*device_handle));
             return result;
         }
     } else {
@@ -226,7 +226,7 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
         if (battery.status == BATTERY_AVAILABLE) {
             result.status = FEATURE_SUCCESS;
             result.value  = battery.level;
-            asprintf(&result.message, "Battery: %d%%", battery.level);
+            _asprintf(&result.message, "Battery: %d%%", battery.level);
         } else if (battery.status == BATTERY_CHARGING) {
             result.status  = FEATURE_INFO;
             result.value   = battery.level;
@@ -244,9 +244,9 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
             result.value  = (int)battery.status;
 
             if (device_found->idProduct != PRODUCT_TESTDEVICE)
-                asprintf(&result.message, "Error retrieving battery status. Error: %ls", hid_error(*device_handle));
+                _asprintf(&result.message, "Error retrieving battery status. Error: %ls", hid_error(*device_handle));
             else // dont call hid_error on test device
-                asprintf(&result.message, "Error retrieving battery status");
+                _asprintf(&result.message, "Error retrieving battery status");
         }
         return result;
     }
@@ -269,7 +269,7 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
         if (ret >= 0) {
             result.status = FEATURE_SUCCESS;
             result.value  = ret;
-            asprintf(&result.message, "Chat-Mix: %d", ret);
+            _asprintf(&result.message, "Chat-Mix: %d", ret);
         } else {
             result.status  = FEATURE_ERROR;
             result.value   = ret;
@@ -324,19 +324,19 @@ static FeatureResult handle_feature(struct device* device_found, hid_device** de
 
     switch (ret) {
     case HSC_READ_TIMEOUT:
-        asprintf(&result.message, "Failed to set/request %s, because of timeout", capabilities_str[cap]);
+        _asprintf(&result.message, "Failed to set/request %s, because of timeout", capabilities_str[cap]);
         break;
     case HSC_ERROR:
-        asprintf(&result.message, "Failed to set/request %s. HeadsetControl Error", capabilities_str[cap]);
+        _asprintf(&result.message, "Failed to set/request %s. HeadsetControl Error", capabilities_str[cap]);
         break;
     case HSC_OUT_OF_BOUNDS:
-        asprintf(&result.message, "Failed to set/request %s. Provided parameter out of boundaries", capabilities_str[cap]);
+        _asprintf(&result.message, "Failed to set/request %s. Provided parameter out of boundaries", capabilities_str[cap]);
         break;
     default: // Must be a HID error
         if (device_found->idProduct != PRODUCT_TESTDEVICE)
-            asprintf(&result.message, "Failed to set/request %s. Error: %d: %ls", capabilities_str[cap], ret, hid_error(*device_handle));
+            _asprintf(&result.message, "Failed to set/request %s. Error: %d: %ls", capabilities_str[cap], ret, hid_error(*device_handle));
         else // dont call hid_error on test device, it will confuse users/devs because it will show success
-            asprintf(&result.message, "Failed to set/request %s. Error: %d", capabilities_str[cap], ret);
+            _asprintf(&result.message, "Failed to set/request %s. Error: %d", capabilities_str[cap], ret);
 
         break;
     }

--- a/src/output.c
+++ b/src/output.c
@@ -155,8 +155,8 @@ HeadsetControlStatus initializeStatus(int num_devices)
 void initializeHeadsetInfo(HeadsetInfo* info, struct device* device)
 {
     info->status = STATUS_SUCCESS;
-    asprintf(&info->idVendor, "0x%04x", device->idVendor);
-    asprintf(&info->idProduct, "0x%04x", device->idProduct);
+    _asprintf(&info->idVendor, "0x%04x", device->idVendor);
+    _asprintf(&info->idProduct, "0x%04x", device->idProduct);
     info->device_name  = device->device_name;
     info->vendor_name  = device->device_hid_vendorname;
     info->product_name = device->device_hid_productname;

--- a/src/utility.c
+++ b/src/utility.c
@@ -198,7 +198,7 @@ fail:
     return (-1);
 }
 
-int asprintf(char** str, const char* fmt, ...)
+int _asprintf(char** str, const char* fmt, ...)
 {
     va_list ap;
     int ret;

--- a/src/utility.h
+++ b/src/utility.h
@@ -94,4 +94,4 @@ int get_float_data_from_parameter(char* input, float* dest, size_t len);
 
 int vasprintf(char** str, const char* fmt, va_list ap);
 
-int asprintf(char** str, const char* fmt, ...);
+int _asprintf(char** str, const char* fmt, ...);

